### PR TITLE
Change vector and tensor return types

### PR DESCRIPTION
**Pull request by lindsayad**
This allows compatibility with automatic differentiation calculations

### CODE REVIEW

**Comment by roystgnr** [comment]
I guess we'll be replacing ```macro_syntax_function()``` with ```// remember not to use a semicolon after this macro```?

File: include/numerics/tensor_tools.h [hunk 1]
```diff
@@ -23,6 +23,10 @@
 // Local includes
 #include "libmesh/libmesh_common.h"
 #include "libmesh/compare_types.h"
+#ifdef LIBMESH_HAVE_METAPHYSICL
+#include "metaphysicl/dualnumber.h"
+#include "metaphysicl/dynamicsparsenumberbase.h"
+#endif
```
Comment (roystgnr) [inline include/numerics/tensor_tools.h]: Is this the right way to do things?  tensor_tools.h is included by both type_vector.h and dense_vector.h so we'd basically be adding a metaphysicl dependency to nearly every file in libMesh and MOOSE!

I know it's a pain to have to include multiple sets of headers (and in the right order, even if we need go so far as to use _decl headers to handle nesting) to do generic programming with their combination, but it's better than having everything depend on everything else - I'd even go so far as to make that norm_sq case take ```const T & a``` so you don't need reference to DualNumber at all there.

**Comment by lindsayad** [comment]
> I guess we'll be replacing `macro_syntax_function()` with `// remember not to use a semicolon after this macro`?

I guess off the top of my head I don't understand why the complex-inf build failed but all the others succeeded?

**Comment by roystgnr** [comment]
That's just the only build (aside from the GRINS builds, which are finally back up and which also failed) where we have -Werror turned on at configure time.

**Comment by lindsayad** [comment]
ah got ya

**Comment by lindsayad** [comment]
Ok, looking like https://github.com/lindsayad/moose/commit/366ee0c97f95efd3c174459ac72324b81f5dafee would about do it

**Comment by pbauman** [comment]
Here's the failure at the "Build libMesh" step of the devel recipe (identical failure on the dbg build as well):
<pre>In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h: In instantiation of ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’:
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
./include/libmesh/type_vector.h:910:55: error: no matching function for call to ‘norm(MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>)’
 auto TypeVector<T>::size() const -> decltype(std::norm(T()))
                                              ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note: candidate: template<class _Tp> _Tp std::norm(const std::complex<_Tp>&)
     norm(const complex<_Tp>& __z)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note:   template argument deduction/substitution failed:
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h:910:55: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const std::complex<_Tp>’
 auto TypeVector<T>::size() const -> decltype(std::norm(T()))
                                              ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note: candidate: template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp)
     norm(_Tp __x)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note:   template argument deduction/substitution failed:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex: In substitution of ‘template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp) [with _Tp = MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>]’:
./include/libmesh/type_vector.h:910:55:   required from ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: error: no type named ‘__type’ in ‘struct __gnu_cxx::__promote<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>, false>’
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h: In instantiation of ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’:
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
./include/libmesh/type_vector.h:1117:6: note: candidate: template<class T> decltype (std::norm(T())) std::norm(const libMesh::TypeVector<T>&)
 auto norm(const libMesh::TypeVector<T> & vector) -> decltype(std::norm(T()))
      ^~~~
./include/libmesh/type_vector.h:1117:6: note:   template argument deduction/substitution failed:
./include/libmesh/type_vector.h:910:55: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const libMesh::TypeVector<T>’
 auto TypeVector<T>::size() const -> decltype(std::norm(T()))
                                              ~~~~~~~~~^~~~~
./include/libmesh/type_vector.h:921:55: error: no matching function for call to ‘norm(MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>)’
 auto TypeVector<T>::norm() const -> decltype(std::norm(T()))
                                              ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note: candidate: template<class _Tp> _Tp std::norm(const std::complex<_Tp>&)
     norm(const complex<_Tp>& __z)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note:   template argument deduction/substitution failed:
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h:921:55: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const std::complex<_Tp>’
 auto TypeVector<T>::norm() const -> decltype(std::norm(T()))
                                              ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note: candidate: template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp)
     norm(_Tp __x)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note:   template argument deduction/substitution failed:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex: In substitution of ‘template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp) [with _Tp = MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>]’:
./include/libmesh/type_vector.h:921:55:   required from ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: error: no type named ‘__type’ in ‘struct __gnu_cxx::__promote<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>, false>’
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h: In instantiation of ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’:
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
./include/libmesh/type_vector.h:1117:6: note: candidate: template<class T> decltype (std::norm(T())) std::norm(const libMesh::TypeVector<T>&)
 auto norm(const libMesh::TypeVector<T> & vector) -> decltype(std::norm(T()))
      ^~~~
./include/libmesh/type_vector.h:1117:6: note:   template argument deduction/substitution failed:
./include/libmesh/type_vector.h:921:55: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const libMesh::TypeVector<T>’
 auto TypeVector<T>::norm() const -> decltype(std::norm(T()))
                                              ~~~~~~~~~^~~~~
./include/libmesh/type_vector.h:941:58: error: no matching function for call to ‘norm(MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>)’
 auto TypeVector<T>::size_sq() const -> decltype(std::norm(T()))
                                                 ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note: candidate: template<class _Tp> _Tp std::norm(const std::complex<_Tp>&)
     norm(const complex<_Tp>& __z)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note:   template argument deduction/substitution failed:
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h:941:58: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const std::complex<_Tp>’
 auto TypeVector<T>::size_sq() const -> decltype(std::norm(T()))
                                                 ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note: candidate: template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp)
     norm(_Tp __x)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note:   template argument deduction/substitution failed:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex: In substitution of ‘template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp) [with _Tp = MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>]’:
./include/libmesh/type_vector.h:941:58:   required from ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: error: no type named ‘__type’ in ‘struct __gnu_cxx::__promote<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>, false>’
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h: In instantiation of ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’:
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
./include/libmesh/type_vector.h:1117:6: note: candidate: template<class T> decltype (std::norm(T())) std::norm(const libMesh::TypeVector<T>&)
 auto norm(const libMesh::TypeVector<T> & vector) -> decltype(std::norm(T()))
      ^~~~
./include/libmesh/type_vector.h:1117:6: note:   template argument deduction/substitution failed:
./include/libmesh/type_vector.h:941:58: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const libMesh::TypeVector<T>’
 auto TypeVector<T>::size_sq() const -> decltype(std::norm(T()))
                                                 ~~~~~~~~~^~~~~
./include/libmesh/type_vector.h:952:58: error: no matching function for call to ‘norm(MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>)’
 auto TypeVector<T>::norm_sq() const -> decltype(std::norm(T()))
                                                 ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note: candidate: template<class _Tp> _Tp std::norm(const std::complex<_Tp>&)
     norm(const complex<_Tp>& __z)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:661:5: note:   template argument deduction/substitution failed:
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h:952:58: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const std::complex<_Tp>’
 auto TypeVector<T>::norm_sq() const -> decltype(std::norm(T()))
                                                 ~~~~~~~~~^~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/core/include/metaphysicl/compare_types.h:32:0,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberbase_decl.h:38,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/contrib/metaphysicl/src/numerics/include/metaphysicl/dynamicsparsenumberarray_decl.h:32,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:33:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note: candidate: template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp)
     norm(_Tp __x)
     ^~~~
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: note:   template argument deduction/substitution failed:
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex: In substitution of ‘template<class _Tp> typename __gnu_cxx::__promote<_Tp>::__type std::norm(_Tp) [with _Tp = MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>]’:
./include/libmesh/type_vector.h:952:58:   required from ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
/femputer/pbauman/swstack/apps/gcc/6.2.0/include/c++/6.2.0/complex:1849:5: error: no type named ‘__type’ in ‘struct __gnu_cxx::__promote<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>, false>’
In file included from ./include/libmesh/point.h:24:0,
                 from ./include/libmesh/dof_map.h:35,
                 from /femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:68:
./include/libmesh/type_vector.h: In instantiation of ‘class libMesh::TypeVector<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’:
./include/libmesh/vector_value.h:41:7:   required from ‘class libMesh::VectorValue<MetaPhysicL::DynamicSparseNumberArray<double, unsigned int> >’
/femputer/pbauman/civet_build_testing_root/libMesh/src/systems/system_projection.C:557:30:   required from here
./include/libmesh/type_vector.h:1117:6: note: candidate: template<class T> decltype (std::norm(T())) std::norm(const libMesh::TypeVector<T>&)
 auto norm(const libMesh::TypeVector<T> & vector) -> decltype(std::norm(T()))
      ^~~~
./include/libmesh/type_vector.h:1117:6: note:   template argument deduction/substitution failed:
./include/libmesh/type_vector.h:952:58: note:   ‘MetaPhysicL::DynamicSparseNumberArray<double, unsigned int>’ is not derived from ‘const libMesh::TypeVector<T>’
 auto TypeVector<T>::norm_sq() const -> decltype(std::norm(T()))
                                                 ~~~~~~~~~^~~~~
make[2]: *** [src/systems/libmesh_devel_la-system_projection.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/femputer/pbauman/civet_build_testing_root/libmesh-devel'
make[1]: *** [../libmesh_devel.la] Error 2
make[1]: Leaving directory `/femputer/pbauman/civet_build_testing_root/libmesh-devel/tests'
make: *** [all-recursive] Error 1
ERROR: exiting with code 2
</pre>

**Comment by lindsayad** [comment]
@roystgnr I think is ready for a new look-over

**Comment by roystgnr** [comment]
Thanks!

**Pull request merged**